### PR TITLE
Java: Add the XRANGE command

### PR DESCRIPTION
**Pull request by acarbonetto**
[XRANGE cmd](https://redis.io/docs/latest/commands/xrange/)

### CODE REVIEW

File: glide-core/src/client/value_conversion.rs [hunk 1]
```diff
@@ -527,7 +527,7 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
 
     // TODO use enum to avoid mistakes
     match command.as_slice() {
-        b"HGETALL" | b"XREAD" | b"CONFIG GET" | b"FT.CONFIG GET" | b"HELLO" => {
+        b"HGETALL" | b"XREAD" | b"CONFIG GET" | b"FT.CONFIG GET" | b"HELLO" | b"XRANGE" => {
             Some(ExpectedReturnType::Map)
```
Comment (tjzhang-BQ) [inline glide-core/src/client/value_conversion.rs:531]: small question, I see that the return type for this cmd is also a map of arrays of strings, why does an ExpectedReturnType of `Map` work here but a new ExpectedType had to be introduced for `LMPop`?

**Review by tjzhang-BQ — APPROVED** [review]

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 2]
```diff
@@ -105,4 +105,29 @@ public interface StreamBaseCommands {
...
     CompletableFuture<Long> xdel(String key, String[] ids);
+
+    /**
+     * Returns stream entries matching a given range of IDs.
+     *
+     * @param key The key of the stream.
+     * @param start Starting ID to search. Use <code>"-"</code> to start with the minimum possible ID.
+     *     Include a <code>"("</code> prior to the ID to do an exclusive search.
+     * @param start End ID to search, or <code>"+"</code> to end with the maximum possible ID. Include
+     *     a <code>"("</code> prior to the ID to do an exclusive search.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:116]: I think `start` and `end` should be an option

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 3]
```diff
@@ -105,4 +105,29 @@ public interface StreamBaseCommands {
...
+     */
+    CompletableFuture<Map<String, String[]>> xrange(String key, String start, String end);
+
+    /**
+     * Returns stream entries matching a given range of IDs.
+     *
+     * @param key The key of the stream.
+     * @param start Starting ID to search. Use <code>"-"</code> to start with the minimum possible ID.
+     *     Include a <code>"("</code> prior to the ID to do an exclusive search.
+     * @param start End ID to search, or <code>"+"</code> to end with the maximum possible ID. Include
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:127]:
```suggestion
     * @param end End ID to search, or <code>"+"</code> to end with the maximum possible ID. Include
```

File: glide-core/src/client/value_conversion.rs (hunk 1 above)
Comment (acarbonetto) [reply glide-core/src/client/value_conversion.rs:531]: I'm not sure why we didn't use `ExpectedReturnType::Map` for LMPOP actually. 
I believe it should return the same result... 
(with the one exception that LMPOP only works with exactly one key - but that's not really important)

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 4]
```diff
@@ -105,4 +105,29 @@ public interface StreamBaseCommands {
...
+     *     a <code>"("</code> prior to the ID to do an exclusive search.
+     * @return A <code>Map</code> of key to stream entry data.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:117]: Can you reword it to be a bit more verbose?

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 5]
```diff
@@ -105,4 +105,29 @@ public interface StreamBaseCommands {
...
+     */
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:118]: TODO example

**Review by yipin-chen — APPROVED** [review]

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 6]
```diff
@@ -3034,6 +3034,64 @@ public void xdel(BaseClient client) {
...
+
+        assertEquals(
+                streamId3,
+                client
+                        .xadd(
+                                key,
+                                Map.of("f3", "foo3", "f4", "bar3"),
+                                StreamAddOptions.builder().id(streamId3).build())
+                        .get());
+
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3080]: Add a test where you do `xrange` after deleting all entries from a stream. Note: don't delete the key, but use `xdel` to make stream empty.

**Review by Yury-Fridlyand — COMMENTED** [review]
Add transaction IT

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 6 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3080]: This is the same as the key2 example... but ok.

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 2 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:116]: I was thinking of doing that... which the options being an ENUM with +/-,  with an optional exclusive search. 
Is that what you're thinking...?

File: glide-core/src/client/value_conversion.rs (hunk 1 above)
Comment (acarbonetto) [reply glide-core/src/client/value_conversion.rs:531]: Confirmed.  
I changed LMPOP value conversion to `Some(ExpectedReturnType::Map)` and it works.  
@tjzhang-BQ you can consider cleaning that up unless you feel strongly that the array.len() == 1 check is important.

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java]: Yes, we have similar thing for `zrange` commands made by San the Great.

File: java/client/src/main/java/glide/api/models/commands/stream/StreamRange.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        public IdBound(String id, boolean isInclusive) {
+            redisApi = isInclusive ? id : "(" + id;
+        }
+    }
+
+    static String[] toArgs(StreamRange start, StreamRange end) {
+        return new String[] {start.getRedisApi(), end.getRedisApi()};
+    }
+
+    static String[] toArgs(StreamRange start, StreamRange end, long count) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamRange.java:92]: Add some docs, because these methods exposed to a user, even though it should not.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 8]
```diff
@@ -548,6 +550,9 @@ private static Object[] streamCommands(BaseTransaction<?> transaction) {
             "0-2", // xadd(streamKey1, Map.of("field2", "value2"), ... .id("0-2").build());
             "0-3", // xadd(streamKey1, Map.of("field3", "value3"), ... .id("0-3").build());
             3L, // xlen(streamKey1)
+            Map.of("0-1", new String[] {"field1", "value1"}), // .xrange(streamKey1, "0-1", "0-1")
+            //            Map.of("0-1", new String[] {"field1", "value1"}), // .xrange(streamKey1, "0-1",
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:554]: todo

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 9]
```diff
@@ -3034,6 +3044,87 @@ public void xdel(BaseClient client) {
...
+                client.xrange(key2, InfRangeBound.MIN, InfRangeBound.MAX).get();
+        assertEquals(0, emptyResult.size());
+
+        assertEquals(OK, client.set(key2, "not_a_stream").get());
+        ExecutionException executionException =
+                assertThrows(
+                        ExecutionException.class,
+                        () -> client.xrange(key2, InfRangeBound.MIN, InfRangeBound.MAX).get());
+        assertInstanceOf(RequestException.class, executionException.getCause());
+
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3120]: Add tests with non min/max ranges too

Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3106]: included test that will fail for non-ID strings

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 10]
```diff
@@ -2658,12 +2660,55 @@ public T xlen(@NonNull String key) {
...
+     *
+     * @param key The key of the stream.
+     * @param start Starting stream ID bound for range, use {@link StreamRange.IdBound#of} to specify
+     *     a stream ID, or {@link StreamRange.IdBound#ofExclusive} to specify an exclusive bounded
+     *     stream ID. Use {@link StreamRange.InfRangeBound#MIN} to start with the minimum available
+     *     ID.
+     * @param end Ending stream ID bound for range, use {@link StreamRange.IdBound#of} to specify a
+     *     stream ID, or {@link StreamRange.IdBound#ofExclusive} to specify an exclusive bounded
+     *     stream ID. Use {@link StreamRange.InfRangeBound#MAX>} to end with the maximum available ID.
+     *     * @return Command Response - A <code>Map</code> of key to stream entry data, where entry
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2680]:
```suggestion
     * @return Command Response - A <code>Map</code> of key to stream entry data, where entry
```

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:2680]: and rerun spotless

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 9 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3106]: I'd like to see a test with `0-1 (5` as a more realistic scenario

**Review by Yury-Fridlyand — APPROVED** [review]
